packages/core-backend/src/BackendWebSocketService.ts

```diff
@@ -0,0 +1,1477 @@
+   * Checks if a message is a server response (correlated with client requests)
+   *
+   * @param message - The message to check
+   * @returns True if the message is a server response
+   */
+  #isServerResponse(message: WebSocketMessage): boolean {
+    return (
+      'data' in message &&
+      message.data &&
+      typeof message.data === 'object' &&
+      'requestId' in message.data
+    );
+  }
+
+  /**
+   * Checks if a message is a subscription notification (has subscriptionId)
+   *
+   * @param message - The message to check
+   * @returns True if the message is a subscription notification with subscriptionId
+   */
+  #isSubscriptionNotification(message: WebSocketMessage): boolean {
+    return (
+      'subscriptionId' in message &&
+      (message as ServerNotificationMessage).subscriptionId !== undefined &&
+      !this.#isServerResponse(message)
+    );
+  }
+
+  /**
+   * Checks if a message has a channel property (system or subscription notification)
+   *
+   * @param message - The message to check
+   * @returns True if the message has a channel property
+   */
+  #isChannelMessage(
+    message: WebSocketMessage,
+  ): message is ServerNotificationMessage {
+    return 'channel' in message;
+  }
+
+  /**
+   * Handles server response messages (correlated with client requests)
+   *
+   * @param message - The server response message to handle
+   */
+  #handleServerResponse(message: ServerResponseMessage): void {
+    const { requestId } = message.data;
+
+    if (!this.#pendingRequests.has(requestId)) {
+      return;
+    }
+
+    const request = this.#pendingRequests.get(requestId);
+    if (!request) {
+      return;
+    }
+
+    this.#pendingRequests.delete(requestId);
+    clearTimeout(request.timeout);
+
+    // Check if the response indicates failure
+    if (message.data.failed && message.data.failed.length > 0) {
+      request.reject(
+        new Error(`Request failed: ${message.data.failed.join(', ')}`),
+      );
+    } else {
+      request.resolve(message.data);
+    }
+  }
+
+  /**
+   * Handles messages with channel properties by triggering channel callbacks
+   *
+   * @param message - The message with channel property to handle
+   */
+  #handleChannelMessage(message: ServerNotificationMessage): void {
+    this.#triggerChannelCallbacks(message);
+  }
+
+  /**
+   * Handles server notifications with subscription IDs
+   *
+   * @param message - The server notification message to handle
+   */
+  #handleSubscriptionNotification(message: ServerNotificationMessage): void {
+    const { subscriptionId } = message;
+
+    // Guard: Only handle if subscriptionId exists
+    if (!subscriptionId) {
+      return;
+    }
+
+    // Fast path: Direct callback routing by subscription ID
+    const subscription = this.#subscriptions.get(subscriptionId);
+    if (subscription) {
+      const { callback } = subscription;
+      // Development: Full error handling
+      if (process.env.NODE_ENV === 'development') {
```

Comment:
We try to keep code in `core` agnostic of the current environment. If some behavior needs to differ in production vs. development, typically we bake customization into the thing we are building. What are your thoughts on adding an option to the constructor that turns on debugging for errors?

Alternatively, in this case, do we need to do something different in production? How much slower is using a `try`/`catch` than not?